src/expr/sequence.cpp

```diff
@@ -0,0 +1,322 @@
+{
+  if (size() < y.size() + start)
+  {
+    return std::string::npos;
+  }
+  if (y.empty())
+  {
+    return start;
+  }
+  if (empty())
+  {
+    return std::string::npos;
+  }
+  std::vector<Node>::const_reverse_iterator itr = std::search(
+      d_seq.rbegin() + start, d_seq.rend(), y.d_seq.rbegin(), y.d_seq.rend());
+  if (itr != d_seq.rend())
+  {
+    return itr - d_seq.rbegin();
+  }
+  return std::string::npos;
+}
+
+bool Sequence::hasPrefix(const Sequence& y) const
+{
+  size_t s = size();
+  size_t ys = y.size();
+  if (ys > s)
+  {
+    return false;
+  }
+  for (size_t i = 0; i < ys; i++)
+  {
+    if (d_seq[i] != y.d_seq[i])
+    {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool Sequence::hasSuffix(const Sequence& y) const
+{
+  size_t s = size();
+  size_t ys = y.size();
+  if (ys > s)
+  {
+    return false;
+  }
+  size_t idiff = s - ys;
+  for (size_t i = 0; i < ys; i++)
+  {
+    if (d_seq[i + idiff] != y.d_seq[i])
+    {
+      return false;
+    }
+  }
+  return true;
+}
+
+Sequence Sequence::replace(const Sequence& s, const Sequence& t) const
+{
+  Assert(d_type == s.d_type);
+  Assert(d_type == t.d_type);
+  size_t ret = find(s);
+  if (ret != std::string::npos)
+  {
+    std::vector<Node> vec;
+    vec.insert(vec.begin(), d_seq.begin(), d_seq.begin() + ret);
+    vec.insert(vec.end(), t.d_seq.begin(), t.d_seq.end());
+    vec.insert(vec.end(), d_seq.begin() + ret + s.size(), d_seq.end());
+    return Sequence(d_type, vec);
+  }
+  return *this;
+}
+
+Sequence Sequence::substr(size_t i) const
+{
+  Assert(i <= size());
+  std::vector<Node> ret_vec;
+  std::vector<Node>::const_iterator itr = d_seq.begin() + i;
+  ret_vec.insert(ret_vec.end(), itr, d_seq.end());
+  return Sequence(d_type, ret_vec);
+}
+
+Sequence Sequence::substr(size_t i, size_t j) const
+{
+  Assert(i + j <= size());
+  std::vector<Node> ret_vec;
+  std::vector<Node>::const_iterator itr = d_seq.begin() + i;
+  ret_vec.insert(ret_vec.end(), itr, itr + j);
+  return Sequence(d_type, ret_vec);
+}
+
+bool Sequence::noOverlapWith(const Sequence& y) const
+{
+  Assert(d_type == y.d_type);
+  return y.find(*this) == std::string::npos
+         && this->find(y) == std::string::npos && this->overlap(y) == 0
+         && y.overlap(*this) == 0;
+}
+
+size_t Sequence::maxSize() { return std::numeric_limits<uint32_t>::max(); }
+
+std::ostream& operator<<(std::ostream& os, const Sequence& s)
+{
+  const std::vector<Node>& vec = s.getVec();
+  std::stringstream ss;
+  if (vec.empty())
+  {
+    ss << "(as seq.empty " << s.getType() << ")";
+  }
+  else
+  {
+    ss << "(seq.++";
+    for (const Node& n : vec)
+    {
+      ss << " " << n;
+    }
+    ss << ")";
+  }
+  return os << ss.str();
+}
+
+namespace strings {
+
+size_t SequenceHashFunction::operator()(const Sequence& s) const
+{
+  size_t ret = 0;
+  const std::vector<Node>& vec = s.getVec();
+  for (const Node& n : vec)
+  {
+    ret = ret + NodeHashFunction()(n);
```

Comment:
It would be better to combine hashes via [fnv1a](https://github.com/CVC4/CVC4/blob/master/src/util/hash.h#L51). Simply adding hashes has some issues, e.g. "AB" and "BA" end up having the same hash.